src/packages/database/postgres/site-license/hook.ts

```diff
@@ -34,218 +38,313 @@ interface License {
+      if (!is_valid_uuid_string(license_id)) {
+        // The site_license is supposed to be a map from uuid's to settings...
+        // We could put some sort of error here in case, though I don't know what
+        // we would do with it.
+        this.dbg.info(`skipping invalid license ${license_id} -- invalid UUID`);
+        continue;
+      }
+      const license = validLicenses.get(license_id);
+      const state = await this.checkLicense({ license, license_id });
+
+      if (state === "valid") {
+        const upgrades: QuotaSetting = this.extractUpgrades(license);
+
+        this.dbg.verbose("computing run quotas...");
+        const run_quota = compute_total_quota(
+          this.project.settings,
+          this.project.users,
+          nextLicense
+        );
+        const run_quota_with_license = compute_total_quota(
+          this.project.settings,
+          this.project.users,
+          {
+            ...nextLicense,
+            ...{ [license_id]: upgrades },
+          }
+        );
+        this.dbg.silly(`run_quota=${JSON.stringify(run_quota)}`);
+        this.dbg.silly(
+          `run_quota_with_license=${JSON.stringify(run_quota_with_license)}`
+        );
+        if (!isEqual(run_quota, run_quota_with_license)) {
+          this.dbg.info(
+            `License "${license_id}" provides an effective upgrade ${JSON.stringify(
+              upgrades
+            )}.`
+          );
+          nextLicense[license_id] = upgrades;
+        } else {
+          this.dbg.info(
+            `Found a valid license "${license_id}", but it provides nothing new so not using it.`
+          );
+        }
+      } else if (state === "expired") {
+        this.dbg.info(`Removing expired license "${license_id}".`);
+        // due to how jsonb_set works, we have to set this to null,
+        // because otherwise an existing license entry continues to exist.
+        nextLicense[license_id] = null;
+      } else {
+        // in all other cases we keep the license around, but not providing any upgrades
+        this.dbg.info(`Disabling license "${license_id}" -- state=${state}`);
+        nextLicense[license_id] = {};
+      }
+    }
+    return nextLicense;
+  }
+
+  /**
+   * get the upgrade provided by a given license
+   */
+  private extractUpgrades(license): QuotaSetting {
+    if (license == null) throw new Error("bug");
+    // Licenses can specify what they do in two distinct ways: upgrades and quota.
+    const upgrades = (license.get("upgrades")?.toJS() ?? {}) as QuotaSetting;
+    if (upgrades == null) {
+      // This is to make typescript happy since QuotaSetting may be null
+      // (though I don't think upgrades ever could be).
+      throw Error("bug");
+    }
+    const quota = license.get("quota");
+    if (quota) {
+      upgrades["quota"] = quota.toJS() as SiteLicenseQuotaSetting;
+    }
+    // remove any zero values to make frontend client code simpler and avoid waste/clutter.
```

Comment:
There's a function in @cocalc/util to do this...   `import { removeNulls } from "@cocalc/util/misc";
`

Reply:
well, I didn't know about this function, but it doesn't do the same. `removeNulls` checks `if (obj[field] != null) {` ... while here it also removes the entry if it is the number zero. I don't want to change this as part of this PR, just keep it as it was before.

Reply:
OK